File: budget-app-frontend/src/services/apiService.ts
import axios from 'axios';
import { 
  LoginDto, 
  RegisterDto, 
  AuthResult, 
  Account, 
  CreateAccountDto, 
  Transaction, 
  CreateTransactionDto, 
  Household, 
  CreateHouseholdDto,
  ImportResult,
  CsvPreviewRow,
  Category,
  BankConnection,
  CreateBankConnectionDto,
  BankConnectionResult,
  ExternalAccount,
  ExternalTransaction,
  ImportExternalTransactionsDto,
  LinkAccountDto,
  Bank
} from '../types';

const API_BASE_URL = process.env.REACT_APP_API_URL || 'http://localhost:5300/api';

class ApiService {
  private token: string | null = null;

  constructor() {
    this.token = localStorage.getItem('token');
    if (this.token) {
      this.setAuthHeader(this.token);
    }
  }

  private setAuthHeader(token: string) {
    axios.defaults.headers.common['Authorization'] = `Bearer ${token}`;
  }

  private removeAuthHeader() {
    delete axios.defaults.headers.common['Authorization'];
  }

  // Auth methods
  async login(loginDto: LoginDto): Promise<AuthResult> {
    try {
      const response = await axios.post(`${API_BASE_URL}/auth/login`, loginDto);
      const { token, user } = response.data;
      
      this.token = token;
      localStorage.setItem('token', token);
      this.setAuthHeader(token);
      
      return { success: true, token, user };
    } catch (error: any) {
      return { 
        success: false, 
        error: error.response?.data?.error || 'Login failed' 
      };
    }
  }

  async register(registerDto: RegisterDto): Promise<AuthResult> {
    try {
      const response = await axios.post(`${API_BASE_URL}/auth/register`, registerDto);
      const { token, user } = response.data;
      
      this.token = token;
      localStorage.setItem('token', token);
      this.setAuthHeader(token);
      
      return { success: true, token, user };
    } catch (error: any) {
      return { 
        success: false, 
        error: error.response?.data?.error || 'Registration failed' 
      };
    }
  }

  logout() {
    this.token = null;
    localStorage.removeItem('token');
    this.removeAuthHeader();
  }

  // Account methods
  async getAccounts(): Promise<Account[]> {
    const response = await axios.get(`${API_BASE_URL}/accounts`);
    return response.data;
  }

  async getAccount(id: number): Promise<Account> {
    const response = await axios.get(`${API_BASE_URL}/accounts/${id}`);
    return response.data;
  }

  async createAccount(accountDto: CreateAccountDto): Promise<Account> {
    const response = await axios.post(`${API_BASE_URL}/accounts`, accountDto);
    return response.data;
  }

  async updateAccount(id: number, accountDto: CreateAccountDto): Promise<Account> {
    const response = await axios.put(`${API_BASE_URL}/accounts/${id}`, accountDto);
    return response.data;
  }

  async deleteAccount(id: number): Promise<void> {
    await axios.delete(`${API_BASE_URL}/accounts/${id}`);
  }

  async getAccountBalance(id: number): Promise<{ accountId: number; currentBalance: number; lastUpdated: string }> {
    const response = await axios.get(`${API_BASE_URL}/accounts/${id}/balance`);
    return response.data;
  }

  // Transaction methods
  async getTransactions(accountId?: number, householdId?: number): Promise<Transaction[]> {
    const params = new URLSearchParams();
    if (accountId) params.append('accountId', accountId.toString());
    if (householdId) params.append('householdId', householdId.toString());
    
    const response = await axios.get(`${API_BASE_URL}/transactions?${params}`);
    return response.data;
  }

  async getTransaction(id: number): Promise<Transaction> {
    const response = await axios.get(`${API_BASE_URL}/transactions/${id}`);
    return response.data;
  }

  async createTransaction(transactionDto: CreateTransactionDto): Promise<Transaction> {
    const response = await axios.post(`${API_BASE_URL}/transactions`, transactionDto);
    return response.data;
  }

  async updateTransaction(id: number, transactionDto: Partial<CreateTransactionDto>): Promise<Transaction> {
    const response = await axios.put(`${API_BASE_URL}/transactions/${id}`, transactionDto);
    return response.data;
  }

  async deleteTransaction(id: number): Promise<void> {
    await axios.delete(`${API_BASE_URL}/transactions/${id}`);
  }

  // Household methods
  async getHouseholds(): Promise<Household[]> {
    const response = await axios.get(`${API_BASE_URL}/households`);
    return response.data;
  }

  async getHousehold(id: number): Promise<Household> {
    const response = await axios.get(`${API_BASE_URL}/households/${id}`);
    return response.data;
  }

  async createHousehold(householdDto: CreateHouseholdDto): Promise<Household> {
    const response = await axios.post(`${API_BASE_URL}/households`, householdDto);
    return response.data;
  }

  async addHouseholdMember(householdId: number, email: string): Promise<void> {
    await axios.post(`${API_BASE_URL}/households/${householdId}/members`, JSON.stringify(email), {
      headers: { 'Content-Type': 'application/json' }
    });
  }

  async removeHouseholdMember(householdId: number, memberUserId: string): Promise<void> {
    await axios.delete(`${API_BASE_URL}/households/${householdId}/members/${memberUserId}`);
  }

  // Import methods
  async previewCsv(file: File): Promise<CsvPreviewRow[]> {
    const formData = new FormData();
    formData.append('file', file);
    
    const response = await axios.post(`${API_BASE_URL}/import/csv/preview`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  }

  async importCsv(file: File, accountId: number): Promise<ImportResult> {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('accountId', accountId.toString());
    
    const response = await axios.post(`${API_BASE_URL}/import/csv`, formData, {
      headers: { 'Content-Type': 'multipart/form-data' }
    });
    return response.data;
  }

  // Categories (assuming we'll add this endpoint later)
  async getCategories(): Promise<Category[]> {
    try {
      const response = await axios.get(`${API_BASE_URL}/categories`);
      return response.data;
    } catch (error) {
      // For now, return the default categories from the types
      return [
        { id: 1, name: 'Boende', color: '#FF6B6B', icon: 'home', scope: 2, createdAt: '' },
        { id: 2, name: 'Mat', color: '#4ECDC4', icon: 'restaurant', scope: 2, createdAt: '' },
        { id: 3, name: 'Transport', color: '#45B7D1', icon: 'directions_car', scope: 2, createdAt: '' },
        { id: 4, name: 'Försäkring', color: '#96CEB4', icon: 'security', scope: 2, createdAt: '' },
        { id: 5, name: 'Barn', color: '#FFEAA7', icon: 'child_care', scope: 2, createdAt: '' },
        { id: 6, name: 'Nöje', color: '#DDA0DD', icon: 'movie', scope: 2, createdAt: '' },
        { id: 7, name: 'Hälsa', color: '#98D8C8', icon: 'local_hospital', scope: 2, createdAt: '' },
        { id: 8, name: 'Inkomst', color: '#81C784', icon: 'account_balance_wallet', scope: 2, createdAt: '' },
        { id: 9, name: 'Övrigt', color: '#B0BEC5', icon: 'category', scope: 2, createdAt: '' }
      ];
    }
  }

  // PSD2/Open Banking methods
  async getBankConnections(): Promise<BankConnection[]> {
    const response = await axios.get(`${API_BASE_URL}/psd2/connections`);
    return response.data;
  }

  async createBankConnection(dto: CreateBankConnectionDto): Promise<BankConnectionResult> {
    const response = await axios.post(`${API_BASE_URL}/psd2/connections`, dto);
    return response.data;
  }

  async completeBankConnection(connectionId: string, authorizationCode: string): Promise<void> {
    await axios.post(`${API_BASE_URL}/psd2/connections/${connectionId}/complete`, authorizationCode, {
      headers: { 'Content-Type': 'application/json' }
    });
  }

  async disconnectBank(connectionId: number): Promise<void> {
    await axios.delete(`${API_BASE_URL}/psd2/connections/${connectionId}`);
  }

  async getExternalAccounts(connectionId: number): Promise<ExternalAccount[]> {
    const response = await axios.get(`${API_BASE_URL}/psd2/connections/${connectionId}/accounts`);
    return response.data;
  }

  async getExternalTransactions(externalAccountId: number, fromDate?: string, toDate?: string): Promise<ExternalTransaction[]> {
    const params = new URLSearchParams();
    if (fromDate) params.append('fromDate', fromDate);
    if (toDate) params.append('toDate', toDate);
    
    const response = await axios.get(`${API_BASE_URL}/psd2/accounts/${externalAccountId}/transactions?${params}`);
    return response.data;
  }

  async importExternalTransactions(dto: ImportExternalTransactionsDto): Promise<ImportResult> {
    const response = await axios.post(`${API_BASE_URL}/psd2/import`, dto);
    return response.data;
  }

  async syncBankConnection(connectionId: number): Promise<void> {
    await axios.post(`${API_BASE_URL}/psd2/connections/${connectionId}/sync`);
  }

  async linkAccount(dto: LinkAccountDto): Promise<void> {
    await axios.post(`${API_BASE_URL}/psd2/accounts/link`, dto);
  }

  async checkConsentExpiry(): Promise<{ hasExpiringConsents: boolean }> {
    const response = await axios.get(`${API_BASE_URL}/psd2/consent-check`);
    return response.data;
  }

  async getAvailableBanks(): Promise<Bank[]> {
    const response = await axios.get(`${API_BASE_URL}/psd2/banks`);
    return response.data;
  }
}

export const apiService = new ApiService();

File: src/BudgetApp.Blazor/Services/ApiService.cs
using BudgetApp.Blazor.Models;
using Microsoft.Extensions.Options;
using System.Text.Json;
using System.Text;

namespace BudgetApp.Blazor.Services;

public interface IApiService
{
    // Auth methods
    Task<AuthResult> LoginAsync(LoginDto loginDto);
    Task<AuthResult> RegisterAsync(RegisterDto registerDto);
    void Logout();

    // Account methods
    Task<List<Account>> GetAccountsAsync();
    Task<Account> GetAccountAsync(int id);
    Task<Account> CreateAccountAsync(CreateAccountDto accountDto);
    Task<Account> UpdateAccountAsync(int id, CreateAccountDto accountDto);
    Task DeleteAccountAsync(int id);

    // Transaction methods
    Task<List<Transaction>> GetTransactionsAsync();
    Task<Transaction> GetTransactionAsync(int id);
    Task<Transaction> CreateTransactionAsync(CreateTransactionDto transactionDto);
    Task<Transaction> UpdateTransactionAsync(int id, CreateTransactionDto transactionDto);
    Task DeleteTransactionAsync(int id);

    // Household methods
    Task<List<Household>> GetHouseholdsAsync();
    Task<Household> GetHouseholdAsync(int id);
    Task<Household> CreateHouseholdAsync(CreateHouseholdDto householdDto);
    Task<Household> UpdateHouseholdAsync(int id, CreateHouseholdDto householdDto);
    Task DeleteHouseholdAsync(int id);

    // Category methods
    Task<List<Category>> GetCategoriesAsync();

    // Import methods
    Task<ImportResult> ImportCsvAsync(Stream fileStream, string fileName);

    // PSD2 methods
    Task<List<BankConnection>> GetBankConnectionsAsync();
    Task<List<Bank>> GetBanksAsync();
    Task<BankConnectionResult> CreateBankConnectionAsync(CreateBankConnectionDto dto);
    Task CompleteBankConnectionAsync(string connectionId, string authorizationCode);
    Task DisconnectBankAsync(int connectionId);
    Task SyncBankConnectionAsync(int connectionId);
    Task<List<ExternalAccount>> GetExternalAccountsAsync(int connectionId);
    Task<List<ExternalTransaction>> GetExternalTransactionsAsync(int externalAccountId);
    Task<ImportResult> ImportExternalTransactionsAsync(ImportExternalTransactionsDto dto);
    Task LinkAccountAsync(LinkAccountDto dto);
    Task<ConsentCheckResult> CheckConsentExpiryAsync();
}

public class ApiService : IApiService
{
    private readonly HttpClient _httpClient;
    private readonly ApiConfiguration _config;
    private readonly JsonSerializerOptions _jsonOptions;
    private string? _token;

    public ApiService(HttpClient httpClient, IOptions<ApiConfiguration> config)
    {
        _httpClient = httpClient;
        _config = config.Value;
        _jsonOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            PropertyNameCaseInsensitive = true
        };

        // Check for existing token
        // Note: In a real application, you'd want to use a more secure storage mechanism
        // This is simplified for the prototype
    }

    private void SetAuthHeader(string token)
    {
        _token = token;
        _httpClient.DefaultRequestHeaders.Authorization = 
            new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
    }

    private void RemoveAuthHeader()
    {
        _token = null;
        _httpClient.DefaultRequestHeaders.Authorization = null;
    }

    private async Task<T> GetAsync<T>(string endpoint)
    {
        var response = await _httpClient.GetAsync($"{_config.BaseUrl}/{endpoint}");
        response.EnsureSuccessStatusCode();
        var json = await response.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<T>(json, _jsonOptions)!;
    }

    private async Task<T> PostAsync<T>(string endpoint, object data)
    {
        var json = JsonSerializer.Serialize(data, _jsonOptions);
        var content = new StringContent(json, Encoding.UTF8, "application/json");
        var response = await _httpClient.PostAsync($"{_config.BaseUrl}/{endpoint}", content);
        response.EnsureSuccessStatusCode();
        var responseJson = await response.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<T>(responseJson, _jsonOptions)!;
    }

    private async Task<T> PutAsync<T>(string endpoint, object data)
    {
        var json = JsonSerializer.Serialize(data, _jsonOptions);
        var content = new StringContent(json, Encoding.UTF8, "application/json");
        var response = await _httpClient.PutAsync($"{_config.BaseUrl}/{endpoint}", content);
        response.EnsureSuccessStatusCode();
        var responseJson = await response.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<T>(responseJson, _jsonOptions)!;
    }

    private async Task DeleteAsync(string endpoint)
    {
        var response = await _httpClient.DeleteAsync($"{_config.BaseUrl}/{endpoint}");
        response.EnsureSuccessStatusCode();
    }

    // Auth methods
    public async Task<AuthResult> LoginAsync(LoginDto loginDto)
    {
        try
        {
            var result = await PostAsync<AuthResult>("auth/login", loginDto);
            if (result.Success && !string.IsNullOrEmpty(result.Token))
            {
                SetAuthHeader(result.Token);
            }
            return result;
        }
        catch (Exception ex)
        {
            return new AuthResult { Success = false, Error = ex.Message };
        }
    }

    public async Task<AuthResult> RegisterAsync(RegisterDto registerDto)
    {
        try
        {
            var result = await PostAsync<AuthResult>("auth/register", registerDto);
            if (result.Success && !string.IsNullOrEmpty(result.Token))
            {
                SetAuthHeader(result.Token);
            }
            return result;
        }
        catch (Exception ex)
        {
            return new AuthResult { Success = false, Error = ex.Message };
        }
    }

    public void Logout()
    {
        RemoveAuthHeader();
    }

    // Account methods
    public async Task<List<Account>> GetAccountsAsync()
    {
        return await GetAsync<List<Account>>("accounts");
    }

    public async Task<Account> GetAccountAsync(int id)
    {
        return await GetAsync<Account>($"accounts/{id}");
    }

    public async Task<Account> CreateAccountAsync(CreateAccountDto accountDto)
    {
        return await PostAsync<Account>("accounts", accountDto);
    }

    public async Task<Account> UpdateAccountAsync(int id, CreateAccountDto accountDto)
    {
        return await PutAsync<Account>($"accounts/{id}", accountDto);
    }

    public async Task DeleteAccountAsync(int id)
    {
        await DeleteAsync($"accounts/{id}");
    }

    // Transaction methods
    public async Task<List<Transaction>> GetTransactionsAsync()
    {
        return await GetAsync<List<Transaction>>("transactions");
    }

    public async Task<Transaction> GetTransactionAsync(int id)
    {
        return await GetAsync<Transaction>($"transactions/{id}");
    }

    public async Task<Transaction> CreateTransactionAsync(CreateTransactionDto transactionDto)
    {
        return await PostAsync<Transaction>("transactions", transactionDto);
    }

    public async Task<Transaction> UpdateTransactionAsync(int id, CreateTransactionDto transactionDto)
    {
        return await PutAsync<Transaction>($"transactions/{id}", transactionDto);
    }

    public async Task DeleteTransactionAsync(int id)
    {
        await DeleteAsync($"transactions/{id}");
    }

    // Household methods
    public async Task<List<Household>> GetHouseholdsAsync()
    {
        return await GetAsync<List<Household>>("households");
    }

    public async Task<Household> GetHouseholdAsync(int id)
    {
        return await GetAsync<Household>($"households/{id}");
    }

    public async Task<Household> CreateHouseholdAsync(CreateHouseholdDto householdDto)
    {
        return await PostAsync<Household>("households", householdDto);
    }

    public async Task<Household> UpdateHouseholdAsync(int id, CreateHouseholdDto householdDto)
    {
        return await PutAsync<Household>($"households/{id}", householdDto);
    }

    public async Task DeleteHouseholdAsync(int id)
    {
        await DeleteAsync($"households/{id}");
    }

    // Category methods
    public async Task<List<Category>> GetCategoriesAsync()
    {
        return await GetAsync<List<Category>>("categories");
    }

    // Import methods
    public async Task<ImportResult> ImportCsvAsync(Stream fileStream, string fileName)
    {
        using var formData = new MultipartFormDataContent();
        var streamContent = new StreamContent(fileStream);
        formData.Add(streamContent, "file", fileName);

        var response = await _httpClient.PostAsync($"{_config.BaseUrl}/import/csv", formData);
        response.EnsureSuccessStatusCode();
        var json = await response.Content.ReadAsStringAsync();
        return JsonSerializer.Deserialize<ImportResult>(json, _jsonOptions)!;
    }

    // PSD2 methods
    public async Task<List<BankConnection>> GetBankConnectionsAsync()
    {
        return await GetAsync<List<BankConnection>>("psd2/connections");
    }

    public async Task<List<Bank>> GetBanksAsync()
    {
        return await GetAsync<List<Bank>>("psd2/banks");
    }

    public async Task<BankConnectionResult> CreateBankConnectionAsync(CreateBankConnectionDto dto)
    {
        return await PostAsync<BankConnectionResult>("psd2/connections", dto);
    }

    public async Task<List<ExternalAccount>> GetExternalAccountsAsync(int connectionId)
    {
        return await GetAsync<List<ExternalAccount>>($"psd2/connections/{connectionId}/accounts");
    }

    public async Task<List<ExternalTransaction>> GetExternalTransactionsAsync(int externalAccountId)
    {
        return await GetAsync<List<ExternalTransaction>>($"psd2/accounts/{externalAccountId}/transactions");
    }

    public async Task<ImportResult> ImportExternalTransactionsAsync(ImportExternalTransactionsDto dto)
    {
        return await PostAsync<ImportResult>("psd2/import", dto);
    }

    public async Task LinkAccountAsync(LinkAccountDto dto)
    {
        await PostAsync<object>("psd2/link", dto);
    }

    public async Task CompleteBankConnectionAsync(string connectionId, string authorizationCode)
    {
        await PostAsync<object>($"psd2/connections/{connectionId}/complete", authorizationCode);
    }

    public async Task DisconnectBankAsync(int connectionId)
    {
        await DeleteAsync($"psd2/connections/{connectionId}");
    }

    public async Task SyncBankConnectionAsync(int connectionId)
    {
        await PostAsync<object>($"psd2/connections/{connectionId}/sync", new { });
    }

    public async Task<ConsentCheckResult> CheckConsentExpiryAsync()
    {
        return await GetAsync<ConsentCheckResult>("psd2/consent-check");
    }
}